x-pack/plugins/fleet/server/tasks/unenroll_inactive_agents_task.ts

```diff
@@ -0,0 +1,199 @@
+        timeout: TIMEOUT,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return this.runTask(taskInstance, core);
+            },
+            cancel: async () => {
+              this.abortController.abort('Task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  public start = async ({ taskManager }: UnenrollInactiveAgentsTaskStartContract) => {
+    if (!taskManager) {
+      this.logger.error('[UnenrollInactiveAgentsTask] Missing required service during start');
+      return;
+    }
+
+    this.wasStarted = true;
+    this.logger.info(`[UnenrollInactiveAgentsTask] Started with interval of [${INTERVAL}]`);
+
+    try {
+      await taskManager.ensureScheduled({
+        id: this.taskId,
+        taskType: TYPE,
+        scope: SCOPE,
+        schedule: {
+          interval: INTERVAL,
+        },
+        state: {},
+        params: { version: VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task UnenrollInactiveAgentsTask, error: ${e.message}`, e);
+    }
+  };
+
+  private get taskId(): string {
+    return `${TYPE}:${VERSION}`;
+  }
+
+  private endRun(msg: string = '') {
+    this.logger.info(`[UnenrollInactiveAgentsTask] runTask ended${msg ? ': ' + msg : ''}`);
+  }
+
+  public async unenrollInactiveAgents(
+    esClient: ElasticsearchClient,
+    soClient: SavedObjectsClientContract
+  ) {
+    this.logger.debug(
+      `[UnenrollInactiveAgentsTask] Fetching agent policies with unenroll_timeout > 0`
+    );
+    // find all agent policies that are not managed and having unenroll_timeout > 0
+    // limit the search to POLICIES_BATCHSIZE at a time and loop until there are no agent policies left
+    const policiesKuery = `${AGENT_POLICY_SAVED_OBJECT_TYPE}.is_managed: false AND ${AGENT_POLICY_SAVED_OBJECT_TYPE}.unenroll_timeout > 0`;
+    let hasMore = true;
+    let page = 1;
+    while (hasMore) {
+      const agentPolicies = await agentPolicyService.list(soClient, {
+        perPage: POLICIES_BATCHSIZE,
+        page: page++,
+        kuery: policiesKuery,
+      });
+      this.logger.debug(
+        `[UnenrollInactiveAgentsTask] Found "${agentPolicies.items.length}" agent policies with unenroll_timeout > 0`
+      );
+      if (!agentPolicies?.items.length) {
+        hasMore = false;
+        this.endRun('Found no policies to process');
+        return;
+      }
+
+      // find inactive agents enrolled on above policies
+      // limit batch size to UNENROLLMENT_BATCHSIZE to avoid scale issues
+      const kuery = `(${AGENTS_PREFIX}.policy_id:${agentPolicies.items
+        .map((policy) => `"${policy.id}"`)
+        .join(' or ')}) and ${AGENTS_PREFIX}.status: inactive`;
+      const res = await getAgentsByKuery(esClient, soClient, {
+        kuery,
+        showInactive: true,
+        page: 1,
+        perPage: UNENROLLMENT_BATCHSIZE,
```

Comment:
I think we could have a counter outside of the policies loop to sum up how many agents were unenrolled, and stop at 1k (so the next job run would continue). Otherwise the task would try to unenroll 1k agents from all policies, which could worst case be a large number.

Reply:
Yes that's a good idea. Should we lower this value if it's too much?

Reply:
1k doesn't sound too much, in scale tests it's quite quick to do actions on 1k agents. Probably we should run scale tests when the feature is ready.